flogger/flogger-system-backend/src/test/kotlin/com/google/common/flogger/backend/system/given/AbstractBackendTestEnv.kt

```diff
@@ -0,0 +1,147 @@
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package com.google.common.flogger.backend.system.given
+
+import com.google.common.flogger.backend.LogData
+import com.google.common.flogger.backend.system.AbstractBackend
+import java.util.logging.Handler
+import java.util.logging.Level
+import java.util.logging.LogRecord
+import java.util.logging.Logger
+
+/**
+ * An instantiatable [AbstractBackend] that remembers the fact
+ * of usage of [ForcingLogger].
+ */
+internal class TestBackend(logger: Logger) : AbstractBackend(logger) {
+
+    /**
+     * Tells whether the last call to [AbstractBackend.log]
+     * used a forcing logger.
+     */
+    internal var wasForcingLoggerUsed = false
+        private set
+
+    /**
+     * This method is not used by [AbstractBackend].
+     *
+     * [LogData] tests are in `SimpleBackendLoggerSpec`.
+     */
+    override fun log(data: LogData) {
+        // no-op
+    }
+
+    /**
+     * This class never tries to format anything in [TestLogger.log],
+     * so this method is not expected to be ever called.
+     */
+    override fun handleError(error: RuntimeException, badData: LogData) {
+        throw UnsupportedOperationException()
+    }
+
+    /**
+     * Explicitly creates a forcing child logger.
+     *
+     * Normally, the forcing logger is obtained from [LogManager][java.util.logging.LogManager].
+     * But in tests, the used [TestLogger] is an explicit subclass of [Logger],
+     * that is not a part of log manager's hierarchy. So, we have to create
+     * an explicit forcing child logger too.
+     */
+    override fun getForcingLogger(parent: Logger): Logger {
+        return ForcingLogger(parent)
+    }
+
+    /**
+     * An explicit forcing logger that notifies the outer [TestBackend]
+     * that it was used.
+     */
+    private inner class ForcingLogger(parent: Logger) :
+        Logger(parent.name + ".__forced__", null) {
+
+        init {
+            setParent(parent)
+        }
+
+        override fun log(record: LogRecord) {
+            wasForcingLoggerUsed = true
+            super.log(record)
+        }
+    }
+}
+
+/**
+ * A Java logger that remembers the captured and published log messages.
+ *
+ * A message is [captured] when it arrives to [Logger.log] method.
+ * Then, if it passes [Logger.log] and arrives to handlers, it is [published].
+ */
+internal class TestLogger(name: String, level: Level) : Logger(name, null) {
```

Comment:
Same as above.

Generally, it's not a good practice to have all similar things called "TestSomething". It's difficult for developers, as they have to choose a proper "TestSomething" each time they write code. And, given that, say, `Logger`s are all of the same interface, they will clash eventually. being mis-imported.